etag/middleware.go

```diff
@@ -14,31 +15,53 @@ const (
 	IfNoneMatchHeaderKey = "If-None-Match"
 )
 
+var (
+	errDoubleWrite = errors.New("Write called twice with ETag Middleware")
+)
+
 type etagResponseWriter struct {
 	http.ResponseWriter
-	req  *http.Request
-	code int
+	req         *http.Request
+	code        int
+	wroteHeader bool
+	wroteBody   bool
 }
 
 func (erw *etagResponseWriter) Write(b []byte) (int, error) {
-	etag := etag(b)
+	etag := calculateEtag(b)
 	erw.Header().Set(ETagHeaderKey, etag)
+
 	if erw.req.Header.Get(IfNoneMatchHeaderKey) == etag {
```

Comment:
Ok, didn't see that. But still, I think the verification should be done here, not when overriding the writer. That is because I'd still like for the result of a `POST` or `PUT` to have its `Etag` returned.